### PR TITLE
Update MySQL InnoDB parameters

### LINKED ISSUE #501: Innodb buffer pool size calculation is wrong
For 3Gi of requested memory, the buffer pool size should be 1536M (1.5Gi). Right now it is 1Gi due to rounding.
https://github.com/presslabs/mysql-operator/blob/6eb89f8d0cfd160424bf762d61459b9891b64ce2/pkg/internal/mysqlcluster/defaults.go#L167-L168

### LINKED ISSUE #502: Consider changing default InnoDB buffer pool sizes
The current formulas, tend to waste memory, so I propose to update with the following:

- `= < 512Gi`  - don't set
- `=<1Gi` - (Memory - 256Mi) * 0.5
- `=<4Gi` - (Memory - 256Mi) * 0.75
- `> 4Gi` - (Memory - 512Mi) * 0.8

It should also take into account `innodb_buffer_pool_instances`. https://dba.stackexchange.com/questions/62411/is-there-a-rule-of-thumb-regarding-the-size-and-number-of-a-buffer-pool-instance.

We should set `innodb_buffer_pool_instances` to `round(requests.limits.cpu)` and also make sure that a instance is no less than 1G. eg: 
```
max(min(ceil(requests.limits.cpu), floor(innodb_buffer_pool_size/1Gi)), 1)
```

Also note that `innodb_buffer_pool_size` should be a multiple of `innodb_buffer_pool_instances` * ` innodb_buffer_pool_chunk_size `
https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_buffer_pool_chunk_size

### PR DESCRIPTION
**Pull request by AMecea**
fixes #501
fixes #502

---
 - [x] I've made sure the [Changelog.md](https://github.com/presslabs/mysql-operator/blob/master/Changelog.md) will remain up-to-date after this PR is merged.

### CODE REVIEW

File: pkg/internal/mysqlcluster/defaults_test.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+var _ = Describe("MySQL defaults", func() {
+	It("humanize should not round the value", func() {
+		q := resource.MustParse("1.5Gi")
+		hq := humanizeSize(q.Value())
+		Expect(hq.String()).To(Equal("1.5G"))
```
Comment (calind) [inline pkg/internal/mysqlcluster/defaults_test.go]: Does MySQL work with `1.5G` or you need to pass `1500M`?

File: Changelog.md [hunk 2]
```diff
@@ -43,11 +43,14 @@ adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
...
  * Change logging: change `cluster` logging field to `key`; normalize logging and more details;
    output Stackdrive compatible format.
+ * Update cron documentation
 ### Removed
 ### Fixed
  * Update and fix e2e tests
  * Fix double date string in bakup path 
  * Copy the nodeSelector as-is in the statefulset (fixes #454)
+ * Fix flakines in ReadOnly cluster condition (fixes #434)
+ * Fix rounding in computing `innodb-buffer-pool-size` (fixes 501)
```
Comment (calind) [inline Changelog.md]:
```suggestion
 * Fix rounding in computing `innodb-buffer-pool-size` (fixes #477 501)
```

File: pkg/internal/mysqlcluster/defaults_test.go (hunk 1 above)
Comment (AMecea) [reply pkg/internal/mysqlcluster/defaults_test.go]: I've updated `humanizeSize` as follows:
 * it outputs values only in megabytes (`M`);
 * for values less than 1M ti will use bytes.

File: pkg/internal/mysqlcluster/defaults_test.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+var _ = Describe("MySQL defaults unit tests", func() {
+	DescribeTable("humanize should not round the value", func(quantity, expect string) {
+		q := resource.MustParse(quantity)
+		hq := humanizeSize(q.Value())
+		Expect(hq.String()).To(Equal(expect))
+	},
+		Entry("grater than G should use M", "1.5Gi", "1536M"),
+		Entry("smaller than G should use M", "321Mi", "321M"),
+		Entry("large values", "212Gi", "217088M"),
+		Entry("Ki convert to M", "1200Ki", "1M"),
```
Comment (calind) [inline pkg/internal/mysqlcluster/defaults_test.go]: This should be `1228800`. `1024Ki` should be `1M`. 

Maybe we should go the extra mile and use all MySQL supported units:
> Sizes are specified using K, M, or G suffix letters to indicate units of KB, MB, or GB.

So the following tests should pass:
```
2Gi -> 2G
1.5Gi -> 1536M
128Mi -> 128M
1.5Mi -> 1536K
1200Ki -> 1200K
1024Ki -> 1M
0.5Ki -> 512
```

File: pkg/internal/mysqlcluster/mysqlcluster_test.go [hunk 4]
```diff
@@ -94,4 +96,42 @@ var _ = Describe("Test MySQL cluster wrapper", func() {
...
+		Entry("zero", "0", "0", "0", "0"),
+		Entry("< 512Mi < 1", "256Mi", "100m", "0", "1"), // innodb-buffer-pool-size not set
+		Entry("< 512Mi > 1", "384Mi", "10", "0", "1"),   // innodb-buffer-pool-size not set
+		Entry("> 512M", "514Mi", "2", "129M", "1"),
+		Entry("< 4G", "3Gi", "2", "2112M", "2"),
+		Entry("= 4G", "4Gi", "2", "2880M", "2"),
+		Entry("> 4G", "4.5Gi", "2", "3435973836", "2"),
+		Entry("< 4G < 1", "3Gi", "100m", "2112M", "1"),
+		Entry("< 8G", "5.5Gi", "6", "4G", "4"),        // (1024 * 5.5 - 512) * 0.8, instances: 4G/1G = 4
+		Entry("> 8G", "12Gi", "2", "9878424780", "2"), // (1024 * 12) * 0.75
```
Comment (calind) [inline pkg/internal/mysqlcluster/mysqlcluster_test.go]:
```suggestion
		Entry("zero", "0", "0", "0", "0"),
		Entry("< 512Mi < 1", "256Mi", "100m", "0", "1"), // innodb-buffer-pool-size not set
		Entry("< 512Mi > 1", "384Mi", "10", "0", "1"),   // innodb-buffer-pool-size not set, innodb-buffer-pool-instances=1
		Entry("> 512M", "514Mi", "2", "129M", "1"), // (512 - 256) * 0.5
		Entry("< 4G", "3Gi", "2", "2112M", "2"), // (1024*3 - 256) * 0,75
		Entry("< 4G < 1", "3Gi", "100m", "2112M", "1"),		
		Entry("= 4G", "4Gi", "2", "2880M", "2"), // (1024*4 - 256) * 0,75
		Entry("> 4G", "4.8Gi", "2", "3501M", "2"), // (1024*4,8 - 256) * 0,75
		Entry("< 8G", "5.5Gi", "6", "4G", "4"), // (1024 * 5,5 - 512) * 0.8, instances: 4G/1G = 4
		Entry("> 8G", "12Gi", "2", "9420M", "2"), // (1024 * 12 - 512) * 0.8
```

**Comment by jianhaiqing** [comment]
```
Note
To avoid potential performance issues, the number of chunks (innodb_buffer_pool_size / innodb_buffer_pool_chunk_size) should not exceed 1000.
```
[innodb-buffer-pool-resize](https://dev.mysql.com/doc/refman/5.7/en/innodb-buffer-pool-resize.html)

what do you know about potential performance issues ? 
@calind

**Pull request merged**